server/middleware/auth_middleware.go

```diff
@@ -0,0 +1,85 @@
+package middleware
+
+import (
+	"strings"
+
+	"github.com/bitcoin-sv/spv-wallet/engine"
+	"github.com/bitcoin-sv/spv-wallet/engine/spverrors"
+	"github.com/bitcoin-sv/spv-wallet/engine/utils"
+	"github.com/bitcoin-sv/spv-wallet/models"
+	"github.com/bitcoin-sv/spv-wallet/server/reqctx"
+	"github.com/gin-gonic/gin"
+)
+
+// AuthMiddleware will check the request for the xPub or AccessKey header
+func AuthMiddleware() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		xPub := strings.TrimSpace(c.GetHeader(models.AuthHeader))
+		authAccessKey := strings.TrimSpace(c.GetHeader(models.AuthAccessKey))
+
+		userContext, err := tryAuth(c, xPub, authAccessKey)
+
+		if err == nil {
+			c.Set(reqctx.UserContextKey, userContext)
+			c.Next()
+		} else {
+			spverrors.AbortWithErrorResponse(c, err, reqctx.Logger(c))
+		}
+	}
+}
+
+func tryAuth(c *gin.Context, xPub, authAccessKey string) (*reqctx.UserContext, error) {
+	if xPub != "" {
+		return authByXPub(c, xPub)
+	}
+	if authAccessKey != "" {
+		return authByAccessKey(c, authAccessKey)
+	}
+	return nil, spverrors.ErrMissingAuthHeader
+}
+
+func authByXPub(c *gin.Context, xPub string) (*reqctx.UserContext, error) {
+	if _, err := utils.ValidateXPub(xPub); err != nil {
+		return nil, spverrors.ErrAuthorization
+	}
+	config := reqctx.AppConfig(c)
```

Comment:
I'm not sure about using those functions in the middleware.
It makes a hidden dependency between middlewares. And config, and logger and engine etc.

Reply:
And it's fine to me.